controller/api/destination/hostport_adaptor.go

```diff
@@ -0,0 +1,257 @@
+	"github.com/linkerd/linkerd2/controller/api/destination/watcher"
+	"github.com/linkerd/linkerd2/controller/k8s"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	logging "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// hostPortAdaptor receives events from a podWatcher and forwards the pod and
+// protocol updates to an endpointProfileTranslator. If required, it subscribes
+// to associated Server updates. Pod updates are only taken into account to the
+// extent they imply a change in its readiness
+type hostPortAdaptor struct {
+	servers    *watcher.ServerWatcher
+	listener   *endpointProfileTranslator
+	ip         string
+	port       uint32
+	address    *watcher.Address
+	endpoint   *pb.WeightedAddr
+	subscribed bool
+	podReady   bool
+
+	enableH2Upgrade     bool
+	controllerNS        string
+	identityTrustDomain string
+	defaultOpaquePorts  map[uint32]struct{}
+
+	k8sAPI      *k8s.API
+	metadataAPI *k8s.MetadataAPI
+	metrics     prometheus.Gauge
+	log         *logging.Entry
+
+	mu sync.Mutex
+}
+
+// hostIPMetrics is a prometheus gauge shared amongst hostPortAdaptor instances
+var hostIPMetrics = promauto.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Name: "host_port_subscribers",
+		Help: "Counter of subscribes to Pod changes for a given hostIP and port",
+	},
+	[]string{"hostIP", "port"},
+)
+
+func newHostPortAdaptor(
+	k8sAPI *k8s.API,
+	metadataAPI *k8s.MetadataAPI,
+	servers *watcher.ServerWatcher,
+	enableH2Upgrade bool,
+	controllerNS string,
+	identityTrustDomain string,
+	defaultOpaquePorts map[uint32]struct{},
+	ip string,
+	port uint32,
+	listener *endpointProfileTranslator,
+	address *watcher.Address,
+	log *logging.Entry,
+) *hostPortAdaptor {
+	log = log.WithField("component", "hostport-adaptor").WithField("ip", ip)
+
+	podReady := isRunningAndReady(address.Pod)
+
+	// if the label map has already been created, it'll get reused
+	metrics := hostIPMetrics.With(prometheus.Labels{
+		"hostIP": ip,
+		"port":   strconv.FormatUint(uint64(port), 10),
+	})
+
+	return &hostPortAdaptor{
+		servers:             servers,
+		listener:            listener,
+		ip:                  ip,
+		port:                port,
+		address:             address,
+		defaultOpaquePorts:  defaultOpaquePorts,
+		podReady:            podReady,
+		enableH2Upgrade:     enableH2Upgrade,
+		controllerNS:        controllerNS,
+		identityTrustDomain: identityTrustDomain,
+		k8sAPI:              k8sAPI,
+		metadataAPI:         metadataAPI,
+		metrics:             metrics,
+		log:                 log,
+	}
+}
+
+func (pt *hostPortAdaptor) Sync() error {
+	pt.mu.Lock()
+	defer pt.mu.Unlock()
+
+	opaquePorts := getAnnotatedOpaquePorts(pt.address.Pod, pt.defaultOpaquePorts)
+	endpoint, err := pt.createEndpoint(*pt.address, opaquePorts)
+	if err != nil {
+		return fmt.Errorf("failed to create endpoint: %w", err)
+	}
+	pt.endpoint = endpoint
+	pt.log.Debugf("Sync for endpoint %s", pt.endpoint)
```

Comment:
nit: will the endpoint have an underlying string representation? Is it better to use [`pt.endpoint.Addr.String()`](https://github.com/linkerd/linkerd2-proxy-api/blob/main/go/destination/destination.pb.go#L305C2-L305C7)?